app/controllers/transfer_packages_controller.rb

```diff
@@ -1,4 +1,59 @@
 class TransferPackagesController < ApplicationController
+  include Concerns::ViewHelper
+
+  helper_method :samples_data
+
+  def new
+    @view_helper = view_helper({ save_back_path: true })
+    @can_update = true
+
+    @transfer_package = TransferPackage.new(sender_institution: @navigation_context.institution)
+    @available_institutions = available_institutions
+  end
+
+  def create
+    params = transfer_package_params
+    params.merge!({
+      sender_institution: @navigation_context.institution,
+    })
+    if id = params.delete(:receiver_institution).presence
+      params[:receiver_institution] = Institution.find(id)
+    end
+
+    samples = Sample
+      .within(@navigation_context.institution)
+      .find_all_by_any_uuid(params[:sample_transfers_attributes].values.map { |sample_transfer| sample_transfer[:sample_uuid] })
+
+    params[:sample_transfers_attributes].transform_values! do |sample_transfer_attributes|
+      uuid = sample_transfer_attributes[:sample_uuid]
+      sample = samples.find { |st| uuid == st.uuid }
+      raise ActiveRecord::RecordNotFound unless sample
+      raise "Can't transfer QC sample" if sample.is_quality_control?
+      raise "User not authorized for transferring sample #{sample.uuid}" unless authorize_resource?(sample, UPDATE_SAMPLE)
+
+      ActionController::Parameters.new(sample: sample).permit!
+    end
+
+    # FIXME: permit! shouldn't be necessary here
+    params.permit!
+
+    @transfer_package = TransferPackage.new(params)
+
+    if @transfer_package.sample_transfers.empty?
+      @transfer_package.errors.add :sample_transfers, "Must not be empty"
+    end
```

Comment:
This could be a validation in the model, then you'd just `if @transfer_package.save` but it means that we can't create a package without sample transfers, which may break the current solution (be doable when we eventually drop the current solution).

Reply:
We'd have to refactor how `TransferPackage#add!` works because you can't save the instance before adding sample transfers, but you need sample transfers to save it. This was just the easiest path to get done.